Neos.RedirectHandler.NeosAdapter/Classes/Service/NodeRedirectionService.php

```diff
@@ -0,0 +1,218 @@
+ *
+ * @Flow\Scope("singleton")
+ */
+class NodeRedirectionService implements NodeRedirectionServiceInterface
+{
+    /**
+     * @Flow\Inject
+     * @var NodeDataRepository
+     */
+    protected $nodeDataRepository;
+
+    /**
+     * @var UriBuilder
+     */
+    protected $uriBuilder;
+
+    /**
+     * @Flow\Inject
+     * @var RedirectionStorageInterface
+     */
+    protected $redirectionStorage;
+
+    /**
+     * @Flow\Inject
+     * @var RouterCachingService
+     */
+    protected $routerCachingService;
+
+    /**
+     * @Flow\Inject
+     * @var PersistenceManagerInterface
+     */
+    protected $persistenceManager;
+
+    /**
+     * @Flow\Inject
+     * @var ContextFactory
+     */
+    protected $contextFactory;
+
+    /**
+     * @Flow\Inject
+     * @var SystemLoggerInterface
+     */
+    protected $systemLogger;
+
+    /**
+     * @Flow\InjectConfiguration(path="statusCode", package="Neos.RedirectHandler")
+     * @var array
+     */
+    protected $defaultStatusCode;
+
+    /**
+     * Creates a redirection for the node if it is a 'TYPO3.Neos:Document' node and its URI has changed
+     *
+     * @param NodeInterface $node The node that is about to be published
+     * @param Workspace $targetWorkspace
+     * @return void
+     * @throws Exception
+     */
+    public function createRedirectionsForPublishedNode(NodeInterface $node, Workspace $targetWorkspace)
+    {
+        $nodeType = $node->getNodeType();
+        if ($targetWorkspace->getName() !== 'live' || !$nodeType->isOfType('TYPO3.Neos:Document')) {
+            return;
+        }
+
+        $context = $this->contextFactory->create([
+            'workspaceName' => 'live',
+            'invisibleContentShown' => true,
+            'dimensions' => $node->getDimensions()
+        ]);
+        $targetNode = $context->getNodeByIdentifier($node->getIdentifier());
+        if ($targetNode === null) {
+            // The page has been added
+            return;
+        }
+
+        $targetNodeUriPath = $this->buildUriPathForNodeContextPath($targetNode->getContextPath());
+        if ($targetNodeUriPath === null) {
+            throw new Exception('The target URI path of the node could not be resolved', 1451945358);
+        }
+
+        $hosts = $this->getHostPatterns($node->getContext());
+
+        // The page has been removed
+        if ($node->isRemoved()) {
+            $this->flushRoutingCacheForNode($targetNode);
+            $statusCode = (integer)$this->defaultStatusCode['gone'];
+            $this->redirectionStorage->addRedirection($targetNodeUriPath, '', $statusCode, $hosts);
+            return;
+        }
+
+        // compare the "old" node URI to the new one
+        $nodeUriPath = $this->buildUriPathForNodeContextPath($node->getContextPath());
+        // use the same regexp than the ContentContextBar Ember View
+        $nodeUriPath = preg_replace('/@[A-Za-z0-9;&,\-_=]+/', '', $nodeUriPath);
```

Comment:
Isn't this because you use beforeNodePublishing? Maybe afterNodePublishing does not have the context information appended? At least I would expect that a redirect would only be created after publishing, not before.